Source/Dafny/Verifier/Translator.BoogieFactory.cs

```diff
@@ -867,5 +867,24 @@ enum BuiltinFunction {
       fvars.Add(BplFormalVar(name, ty, incoming, out var e));
       return e;
     }
+
+    public static IToken ToDafnyToken(Bpl.IToken exprTok) {
+      if (exprTok == null) {
+        return null;
+      } else if (exprTok is IToken t) {
+        return t;
+      } else {
```

Comment:
We have no guarantee that all the Boogie tokens are Dafny tokens. There exists at least Boogie.Token.NoToken which is not a Dafny token, and we don't have a formal contract indicating it will never be passed here.

Note that the code here won't have any performance penalty.

On the other side, I would block a PR that would have the same performance but casts the token without checking:
```
      if (exprTok == null) {
        return null;
      } else {
        return (IToken)t; // I don't see a reason why it shouldn't be a Dafny token
      }
```
We are not writing in Dafny, so robustness implies we take into account potential crashes.
If you prefer, I could write a unit test so that this branch is covered and this case occur.